src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityContext.cs

```diff
@@ -84,9 +85,29 @@ internal sealed class ImmutabilityContext {
 				return true;
 			}
 
-			if( TryGetImmutableTypeInfo( type, out ImmutableTypeInfo info ) ) {
-				if( info.Kind.HasFlag( kind ) ) {
-					return true;
+			if( type is INamedTypeSymbol namedType ) {
+				if( TryGetImmutableTypeInfo( namedType, out ImmutableTypeInfo info ) ) {
+					if( info.Kind.HasFlag( kind ) ) {
+						if( namedType.TypeArguments.Length != info.ImmutableTypeParameters.Length ) {
+							// could this happen?
+							throw new Exception();
+						}
+
+						IEnumerable<(ITypeSymbol, bool)> pairs = namedType
+							.TypeArguments
+							.Zip( info.ImmutableTypeParameters, ( a, relevant ) => (a, relevant) );
+						foreach( (ITypeSymbol argument, bool isRelevant) in pairs ) {
+							if( !isRelevant ) {

```

Comment:
For a "container" type

```csharp
[Immutable] class MyType<[Immutable] A, B> {}
```

Which might get used like

```csharp
[Immutable]
class Holder {
  readonly MyType<int, object> m_held;
}
```

Only `A`/`int` is relevant when considering where it is immutable (this generally implies `B`/`object` isn't "stored").


Also note that for the current analyzers, `MyType<object, object>` is not allowed since `<[Immutable] A` means `A` is _required_ to be immutable. For the new analyzers, `MyType<object, object>` is just not considered to be immutable, where as `MyType<int, object>` is.